So awesome add-ons controller (#592)
kyma-project/test-infra: development/prow-addons-ctrl-manager/README.md

# Prow Addons Controller Manager

## Overview

> The Prow Addons Controller Manager contains all custom controller extension for Prow infrastructure. This project is bootstrapped by Kubebuilder. Check the official [Kubebuilder](https://github.com/kubernetes-sigs/kubebuilder/tree/v1.0.7) documentation to find out how to work with the project. 

Reviewer: ```suggestion
The Prow Addons Controller Manager contains all custom controller extensions for Prow infrastructure. This project is bootstrapped by Kubebuilder. Check the official [Kubebuilder](https://github.com/kubernetes-sigs/kubebuilder/tree/v1.0.7) documentation to find out how to work with the project. 
```

  ...
```

## Usage

### Run a local version

To run the controller outside the cluster, run this command:

```bash
make run
```

### Build a production version

To build the production Docker image, run this command:

```bash
IMG={image_name}:{image_tag} make docker-build
```

The variables are:

* `{image_name}` which is the name of the output image.
* `{image_tag}` which is the tag of the output image.

### Use environment variables
Use the following environment variables to configure the application:

| Name | Required | Default | Description |
|-----|---------|--------|------------|
| **NOTIFIER_GITHUB_ACCESS_TOKEN** | Yes |  | The GitHub token for querying GitHub api. |
| **NOTIFIER_SLACK_TOKEN** | Yes | | The Slack token used or publishing messages to Slack channel. Find more information [here](https://api.slack.com/docs/token-types#bot). |
| **NOTIFIER_SLACK_REPORTER_CHANNEL** | Yes |  | The Slack channel name where notification is posted. You must specify a public channel, private channel, or an IM channel, e.g. for the public channel pass the channel's name (#general). Find more information [here](https://api.slack.com/methods/chat.postMessage#channels).|

> | **NOTIFIER_SLACK_REPORTER_ACT_ON_PROW_JOB_TYPE** | No | `periodic;postsubmit` | The names of the ProwJob types you want to observe. Multiple type names should be separated by comma or semicolon. Find the acceptable job types [here](https://github.com/kubernetes/test-infra/blob/fbc4040f1824bfa126f873650848396a10f05e8a/prow/apis/prowjobs/v1/types.go#L33-L43).

Reviewer: ```suggestion
| **NOTIFIER_SLACK_REPORTER_ACT_ON_PROW_JOB_TYPE** | No | `periodic;postsubmit` | The names of the ProwJob types you want to observe. Separate multiple type names by a comma or a semicolon. Find the acceptable job types [here](https://github.com/kubernetes/test-infra/blob/fbc4040f1824bfa126f873650848396a10f05e8a/prow/apis/prowjobs/v1/types.go#L33-L43).
```

| **NOTIFIER_SLACK_REPORTER_ACT_ON_PROW_JOB_STATE** | No | `failure;error` | The names of the ProwJob states you want to observe. Multiple state names should be separated by comma or semicolon.Find the acceptable job states [here](https://github.com/kubernetes/test-infra/blob/fbc4040f1824bfa126f873650848396a10f05e8a/prow/apis/prowjobs/v1/types.go#L48-L62). 
| **NOTIFIER_SLACK_REPORTER_USER_ICON_EMOJI** | No | `:prow:`  | The Slack bot user name.

> | **NOTIFIER_SLACK_REPORTER_USERNAME** | No | `prow-notifier` | The Emoji to use as the icon for Slack notification message.

Reviewer: ```suggestion
| **NOTIFIER_SLACK_REPORTER_USERNAME** | No | `prow-notifier` | The Emoji to use as the icon for a Slack notification message.
```

### Project structure

<!-- Update the repository structure each time you modify it. -->

The `prow-addons-ctrl-manager` project has the following structure:

```
  .

>   ├── Dockerfile                # Definition used to build the Docker image

Reviewer: ```suggestion
  ├── Dockerfile                # The definition used to build the Docker image
```

>   ├── Makefile                  # Contains all commands required to work with the project, such as test or deploy

Reviewer: ```suggestion
  ├── Makefile                  # All commands required to work with the project, such as "test" or "deploy"
```

>   ├── bin                       # Destination for project compiled binaries  

Reviewer: ```suggestion
  ├── bin                       # The destination for the project's compiled binaries  
```

>   ├── cmd                       # Entrypoint for main application for this project

Reviewer: ```suggestion
  ├── cmd                       # The entrypoint for the main application of this project
```

  ├── config

>   │   ├── crds                  # CRDs definition used in this project. DO NOT EDIT

Reviewer: ```suggestion
  │   ├── crds                  # The CRDs definition used in this project. Do not edit it.
```

>   │   ├── default               # Default patches and resources for manager.yaml 

Reviewer: ```suggestion
  │   ├── default               # Default patches and resources for "manager.yaml"
```

>   │   ├── deploy                # Combined set of resources generated by kustomize tool  

Reviewer: ```suggestion
  │   ├── deploy                # A combined set of resources generated by the "kustomize" tool  
```

>   │   ├── kustomization.yaml    # Instructions for the kustomize tool

Reviewer: ```suggestion
  │   ├── kustomization.yaml    # Instructions for the "kustomize" tool
```

>   │   ├── manager               # Basic resourced for Controller Manager, such as StatefulSet and ServiceAccount 

Reviewer: ```suggestion
  │   ├── manager               # Basic resources for the Controller Manager, such as StatefulSet and ServiceAccount 
```

>   │   └── rbac                  # RBAC resources autogenerate by kustomize tool based on annotation in source code. DO NOT EDIT

Reviewer: ```suggestion
  │   └── rbac                  # RBAC resources autogenerated by the "kustomize" tool based on an annotation in the source code. Do not edit them.
```

  ├── docs                      # Documentation for the project architecture

>   ├── hack                      # Non-build related scripts

Reviewer: ```suggestion
  ├── hack                      # Non-build-related scripts
```

>   └── pkg                       # Contains the controller source code

Reviewer: ```suggestion
  └── pkg                       # The controller source code
```

# Prow Addons Controller Manager

## Overview

The Prow Addons Controller Manager embeds all custom controller extension for Prow infrastructure. This project is bootstrapped by Kubebuilder. Check the official [Kubebuilder](https://github.com/kubernetes-sigs/kubebuilder/tree/v1.0.7) documentation to find out how to work with the project. 

## Prerequisites

Use the following tools to set up the project:

* Version 1.11 or higher of [Go](https://golang.org/dl/)
* Version 0.5.1 or higher of [Dep](https://github.com/golang/dep)
* Version 2.0.0 of [Kustomize](https://github.com/kubernetes-sigs/kustomize)
* Version 1.0.7 of [Kubebuilder](https://github.com/kubernetes-sigs/kubebuilder)
* The latest version of [Docker](https://www.docker.com/)
* The latest version of [Mockery](https://github.com/vektra/mockery) 

## Available controllers

### Notifier
Notifier is responsible for watching all ProwJobs and send notification on Slack channel:

![](./docs/assets/slack_alert.png)

For information about the available configuration, see the **Use environment variables** section. 

> The ProwJob can be ignored by adding label **prow.k8s.io/slack.skipReport** with value `true` under the ProwJob **metadata** entry. 

Author: I was also considering that, but I wanted to make it's as similar as possible to have a simple donate to Prow in the future.  But maybe you right, so, for now, I changed it to `kyma-project.io`.

```yaml
apiVersion: prow.k8s.io/v1
kind: ProwJob
metadata:
  labels:
    prow.k8s.io/slack.skipReport: true # this job will be ignored by Slack reporter
  name: 47291cd0-2bb4-11e9-9e45-0a580a2c0027
  namespace: default
spec:
  job: post-master-kyma-gke-upgrade
  type: postsubmit
  ...

> ```

Author: more details? already we have in doc such sentence
> For information about the available configuration, see the **Use environment variables** section. 

and in **Use environment variables** section there are all possible envs, e.g. `NOTIFIER_SLACK_REPORTER_ACT_ON_PROW_JOB_TYPE` and `NOTIFIER_SLACK_REPORTER_ACT_ON_PROW_JOB_STATE`

which are responsible for configuring alerts for particular ProwJobs

> ```

Reviewer 2: We should show present job configuration, not effective ProwJob IMO.

## Usage

### Run a local version

To run the controller outside the cluster, run this command:

```bash
make run
```

### Build a production version

To build the production Docker image, run this command:

```bash
IMG={image_name}:{image_tag} make docker-build
```

The variables are:

* `{image_name}` which is the name of the output image.
* `{image_tag}` which is the tag of the output image.

### Use environment variables
Use the following environment variables to configure the application:

| Name | Required | Default | Description |
|-----|---------|--------|------------|

> | **NOTIFIER_GITHUB_ACCESS_TOKEN** | Yes |  | The GitHub token for querying GitHub api. |

Author: Currently, the controller-managers are used mostly to aggregate all controllers. There is no option to disable some of them. To do that we need to implement own strategy. IMO for now it's an overhead to have sth like that.

> | **NOTIFIER_GITHUB_ACCESS_TOKEN** | Yes |  | The GitHub token for querying GitHub api. |

Reviewer 2: 👍

| **NOTIFIER_SLACK_TOKEN** | Yes | | The Slack token used or publishing messages to Slack channel. Find more information [here](https://api.slack.com/docs/token-types#bot). |

> | **NOTIFIER_SLACK_REPORTER_CHANNEL** | Yes |  | The Slack channel name where notification are posted. |

Author: done

> | **NOTIFIER_SLACK_REPORTER_ACT_ON_PROW_JOB_TYPE** | No | `periodic;postsubmit` | The names of the ProwJob types you want to observe. Multiple type names should be separated by comma or semicolon.

Author: done

| **NOTIFIER_SLACK_REPORTER_ACT_ON_PROW_JOB_STATE** | No | `failure;error` | The names of the ProwJob states you want to observe. Multiple state names should be separated by comma or semicolon. 

## Development

> 

Author: the `make test` which is executing:
- generate (regenerate all go code, also mocks)
- manifests (create manifests using kustomize)
- executes: fmt, vet, dep status, go test ./pkg ./cmd

previously in doc we have:

> ### Run tests
> 
> To run all unit tests, use the following command:
> 
> ```bash
> make test
> ```
> 

now i change it to:

> ### Run tests
> 
> To test your changes before each commit, use the following command:
> 
> ```bash
> make test
> ```
>
